fe/fe-core/src/main/java/com/starrocks/alter/LakeTableAlterJobV2Builder.java

```diff
@@ -2,12 +2,78 @@
 
 package com.starrocks.alter;
 
-import com.starrocks.common.NotImplementedException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.staros.proto.ShardStorageInfo;
+import com.starrocks.catalog.Column;
+import com.starrocks.catalog.MaterializedIndex;
+import com.starrocks.catalog.Partition;
+import com.starrocks.catalog.Tablet;
+import com.starrocks.catalog.TabletMeta;
+import com.starrocks.common.DdlException;
 import com.starrocks.common.UserException;
+import com.starrocks.lake.LakeTable;
+import com.starrocks.lake.LakeTablet;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.thrift.TStorageMedium;
+
+import java.util.List;
+import java.util.Map;
 
 public class LakeTableAlterJobV2Builder extends AlterJobV2Builder {
+    private final LakeTable table;
+
+    public LakeTableAlterJobV2Builder(LakeTable table) {
+        this.table = table;
+    }
+
     @Override
     public AlterJobV2 build() throws UserException {
-        throw new NotImplementedException("does not support alter lake table yet");
+        if (newIndexSchema.isEmpty() && !hasIndexChanged) {
+            throw new DdlException("Nothing is changed. please check your alter stmt.");
+        }
+        GlobalStateMgr globalStateMgr = GlobalStateMgr.getCurrentState();
+
+        long tableId = table.getId();
+        LakeTableSchemaChangeJob schemaChangeJob = new LakeTableSchemaChangeJob(jobId, dbId, tableId, table.getName(), timeoutMs);
+        schemaChangeJob.setBloomFilterInfo(bloomFilterColumnsChanged, bloomFilterColumns, bloomFilterFpp);
+        schemaChangeJob.setAlterIndexInfo(hasIndexChanged, indexes);
+        schemaChangeJob.setStartTime(startTime);
+        schemaChangeJob.setStorageFormat(newStorageFormat);
+        for (Map.Entry<Long, List<Column>> entry : newIndexSchema.entrySet()) {
+            long originIndexId = entry.getKey();
+            // 1. get new schema version/schema version hash, short key column count
+            String newIndexName = SchemaChangeHandler.SHADOW_NAME_PRFIX + table.getIndexNameById(originIndexId);
+            short newShortKeyColumnCount = newIndexShortKeyCount.get(originIndexId);
+            long shadowIndexId = globalStateMgr.getNextId();
+
+            // create SHADOW index for each partition
+            for (Partition partition : table.getPartitions()) {
+                long partitionId = partition.getId();
+                TStorageMedium medium = table.getPartitionInfo().getDataProperty(partitionId).getStorageMedium();
+
+                List<Tablet> originTablets = partition.getIndex(originIndexId).getTablets();
+                List<Long> shadowTabletIds = createShards(originTablets.size(), table.getShardStorageInfo());
```

Comment:
`table.getShardStorageInfo()` returns table default storage info, include cache info, 
but partitions may have different cache infos from table, 
so we should copy the table default storage info and overwrite the cache info with partition cache info.

Reply:
How can I get partition specific storage info?

Reply:
from `PartitionInfo`.

```
PartitionInfo partitionInfo = table.getPartitionInfo();
StorageInfo partitionStorageInfo = partitionInfo.getStorageInfo(partitionId);
CacheInfo cacheInfo = CacheInfo.newBuilder().setEnableCache(partitionStorageInfo.isEnableStorageCache())
                .setTtlSeconds(partitionStorageInfo.getStorageCacheTtlS())
                .setAllowAsyncWriteBack(partitionStorageInfo.isAllowAsyncWriteBack())
                .build();
ShardStorageInfo shardStorageInfo = ShardStorageInfo.newBuilder(table.getShardStorageInfo())
                .setCacheInfo(cacheInfo).build();
```

You can add a new method in `LakeTable` for this.

Reply:
updated